### DISCUSSION TITLE
SOLVED: Docstring inheritance

### DISCUSSION THREAD

**Discussion by keckler** [terrapower/armi#423] [Ideas]
There are multiple instances of methods that do not have docstrings, and in that case Sphinx will inherit the parent's docsting, if it exists. In some cases, this leads to misleading or inaccurate docstrings which are hard to identify unless one explicitly compares the documentation with the method source. Some particular examples of this are:
- https://terrapower.github.io/armi/.apidocs/armi.materials.zincOxide.html#armi.materials.zincOxide.ZnO.density
- https://terrapower.github.io/armi/.apidocs/armi.materials.yttriumOxide.html#armi.materials.yttriumOxide.Y2O3.density
- https://terrapower.github.io/armi/.apidocs/armi.materials.scandiumOxide.html

I understand that all those material classes are pretty niche, but those just happened to be some of the places that I noticed. The point is that this might be happening all over the codebase, and it is hard to immediately tell.

@john-science and I were discussing how this could be fixed. The most obvious solution is just to add docstrings for all the methods that are missing them. But another option that came up would be to disable to inheritance of docstrings from parent classes within Sphinx, as in something akin to: https://stackoverflow.com/questions/16518834/avoid-docstrings-from-parent-in-sphinx

The disabling of docstring inheritance would fix the described problem across the entire codebase with minimal effort. It would also have the added benefit of making it apparent when reading the documentation which methods are lacking docstrings entirely.

However, there could be undesirable consequences to disabling docstring inheritance, such as wiping out docstrings that indeed _should_ be inherited. Or possibly leaving the codebase with many methods that have no descriptions at all, which could make the documentation possibly less useful.

Does anybody have any thoughts the possibility of disabling docstring inheritance within Sphinx?

@ntouran @jakehader @youngmit

**Comment by jakehader** [comment]
I like the idea of removing docstring inheritance on sphinx. @keckler, @john-science is there a way to configure the linter to tell us how many and which methods don't have docstrings implemented? This might be a good thing to have as part of the code review checklist.

**Comment by keckler** [comment]
Well actually Pylint _does_ check for methods without docstrings and is supposed to warn us. It should even warn in the case that a docstring is inherited, as long as the method itself does not have its own docstring defined. I didn't realize this before, but the issue (at least with the examples that I cited in my original post) is that we have enabled the `docstring-min-length` option in our [`pylintrc`](https://github.com/terrapower/armi/blob/master/pylintrc#L51) so that the docstring check is not performed on any methods with less than 5 lines. 

So I guess the problem really only exists for short methods which have inherited docstrings. 

Not sure what the original reason was for implementing `docstring-min-length`, but maybe we want to do away with that?

**Reply by jakehader** [reply]
We should probably just disable that. Any objections @john-science?

**Reply by john-science** [reply]
Please do not disable that.

The Sphinx settings for docstrings are set the way they are because enabling that feature throws 10,000 warnings at build time. And the vast majority of those warnings are for numpy, scipy, and other third-party libraries that we can't control.

If you want more docstrings, great. Change the Sphinx-docstring-length command as a way to TEST which methods need docstrings. But as of the last time I tried this, we can't have that be the default behavior in trunk.

**Reply by jakehader** [reply]
Good points @john-science. Maybe we can use a different linter (e.g., https://github.com/terrencepreilly/darglint) specifically for checking docstring compliance throughout without sphinx providing the 1000s of warnings. Or if there isn't a better tool than sphinx we can just modify the pylintrc file locally for updating the docstrings and then make sure it's not modified for the PR.

**Comment by keckler** [comment]
To follow up on this dormant thread, here is another example of where docstring inheritance in ARMI becomes super annoying.

As you can see in the image below, all these attributes have the same docstring. And that docstring doesn't even appear to make much sense. Turns out that is a docstring propagating through from Yamlize, and it is all over our API docs (pretty much anytime a class has attributes that are read from user input). So yeah, kinda cluttered and annoying...
![image](https://user-images.githubusercontent.com/21959600/163054703-d2fbeb85-9a47-46c5-8b17-d007616756f9.png)

**Reply by john-science** [reply]
And here:

https://terrapower.github.io/armi/.apidocs/armi.reactor.blueprints.assemblyBlueprint.html#armi.reactor.blueprints.assemblyBlueprint.AssemblyBlueprint.radialMeshPoints

**Comment by ntouran** [comment]
* The traditional docstring linter is https://github.com/pycqa/pydocstyle
* We almost definitely can disable linting for code that is outside of our main tree (e.g. numpy)
* This is a problem for sure that we should fix. the examples just above are pretty bad.

**Reply by john-science** [reply]
I have been playing with this, this morning, and can't find the right flag to turn off the inherited docstrings.  I'll keep playing with it.

We just have so many Sphinx extensions, I  probably just need to find the _one_ that is last in the chain.

**Reply by john-science** [reply]
Okay, I have managed to fix everything except the `Yamlize` objects.  Progress.

**Reply by john-science** [reply]
PR is up:   https://github.com/terrapower/armi/pull/628

**Comment by john-science** [comment]
Okay, the problem is generally fixed in the docs except for subclasses of `yamlize.Attribute`, which I have opened a ticket for here:

https://github.com/terrapower/armi/issues/629

**Reply by john-science** [reply]
An example of a fixed doc (zincOxide):

https://terrapower.github.io/armi/.apidocs/armi.materials.zincOxide.html
